ear-production-suite/plugins/scene/src/scene_frontend_connector.cpp

```diff
@@ -439,6 +439,16 @@ void JuceSceneFrontendConnector::removeFromProgrammes(
       }
     }
   }
+  else {
+    auto* prog = p_->getProgrammeStore()->mutable_programme(p_->getProgrammeStore()->selected_programme_index());
+    for(int elementPos = 0; elementPos < prog->element().size(); ++elementPos) {
+        if(communication::ConnectionId(prog->element(elementPos).object().connection_id()) == id) {
+            prog->mutable_element()->erase(prog->mutable_element()->begin() + elementPos);
```

Comment:
Looking at the code above, the call to removeElement does something similar but locks access to the store first. I'm not sure if that's needed here (if the GUIs inactive), but unless you're sure there's nothing else that might concurrently modify the store via an nng callback or something, might be worth changing this to call that function instead?

Reply:
You're right, it should definitely be locked. I think i cannot use removeElement(int paramIdx, int), because i don't know the parameter index, only the connection ID.

Reply:
I think it's the programme index which you need, which you get via selected_programme_index() a little earlier, although possibly, you might want to iterate through all the programmes if you're pruning dead objects as it might have been active in many?

Reply:
Yes, i just realised that it should be removed from all programs. 
I didn't mean to say parameter index, but element index. But i can calculate it the same way it was originally done, only searching through the programme store and not through the programme container UI element. I think this should be the default behaviour anyway, because it's not relying on an UI element.
```
void JuceSceneFrontendConnector::removeFromProgrammes(
    communication::ConnectionId id) {
  auto* progStore = p_->getProgrammeStore();
  for (int programmeIndex = 0;
       programmeIndex < progStore->programme_size();
       ++programmeIndex) {
    auto elements = progStore->programme(programmeIndex).element();
    auto element =
        std::find_if(elements.begin(), elements.end(), [id](auto entry) {
            return communication::ConnectionId(
                      entry.object().connection_id()) == id;
        });
    if (element != elements.end()) {
      auto elementIndex = std::distance(elements.begin(), element);
      removeElement(programmeIndex, elementIndex);
      
      std::lock_guard<std::mutex> programmeViewsLock(programmeViewsMutex_);
      updateElementOverview(programmeIndex);
    }
  }
}
```
To guarantee that the calculated indices will definitely be valid, we'd need to lock the programme store at the beginning of this scope. But then removeElement() can't obtain the lock. As far as i can see, removeElement() is only called from this and one other place. So the responsibility for locking the programme store could be moved to the caller, which would resolve the issue for now, but would make it more vulnerable for future mistakes.
We could also just copy everything but the locking from removeElement() and inline it here, and lock at the beginning of removeFromProgrammes().